### ISSUE TITLE
Monorepo

### ISSUE THREAD

**Issue by gmaclennan** [digidem/mapeo-core#56]
I've been thinking about whether combining Mapeo into a monorepo would make development and maintenance easier.

We have several modules which are inter-dependent and changes often need to be made to all of them at the same time. An initial list of modules which might be in the monorepo are:

- [`mapeo-core`](https://github.com/digidem/mapeo-core)
- [`mapeo-desktop`](https://github.com/digidem/mapeo-desktop)
- [`mapeo-mobile`](https://github.com/digidem/mapeo-mobile)
- [`mapeo-web`](https://github.com/digidem/mapeo-web)
- [`kappa-osm`](https://github.com/digidem/kappa-osm)
- [`mapeo-settings-builder`](https://github.com/digidem/mapeo-settings-builder)
- [`react-mapfilter`](https://github.com/digidem/react-mapfilter)
- [`mapeo-schema`](https://github.com/digidem/mapeo-schema)
- [`mapeo-server`](https://github.com/digidem/mapeo-server)
- [`osm-p2p-server`](https://github.com/digidem/osm-p2p-server)
- [`id-mapeo`](https://github.com/digidem/id-mapeo)

All of these modules are very closely tied to each other and I've often found that an update requires updating and publishing several of them at the same time, mainly just to bump a dep version. This is what I understand monorepos help solve - the ability to publish inter-dependent modules together. The other advantage is moving issues into the same repo - I have often ended up searching through repos for an issue, and for end-users an issue they experience could already be reported in any of the upstream modules.

The other big reason I've been thinking about monorepos is because I would like to share more code between mapeo-desktop and mapeo-mobile. Currently this is very hard/frustrating to develop on - it requires `npm link` which never seems to work reliably in recent npm versions, and causes all kinds of problems with different versions of react and other shared deps being imported. Publishing when sharing code also requires careful coordination of releases and testing.

I know that monorepos add extra complexity to the dev setup, although I think tooling has improved a lot over the past year or so. At the moment I'm leaning towards the benefits out-weighing the negatives, and am interested in hearing thoughts @karissa @noffle and gathering a list of pros & cons.

How to set this up? [Yarn Workspaces](https://yarnpkg.com/lang/en/docs/workspaces/) allow you to coordinate installs of several packages in the same repo. It can be used on its own to manage a monorepo. [Lerna](https://lerna.js.org) adds a layer of version management and coordinated publishing npm. We would need to adjust our build and testing CI so that it worked within the monorepo structure.

A few packages which use monorepos:

- https://github.com/ReactTraining/react-router/
- https://github.com/mui-org/material-ui

**Comment by gmaclennan** [comment]
An additional thought: having mapeo-mobile and mapeo-desktop in the same repo makes it simpler to setup CI tests which test sync between desktop and mobile, something that I think we need to do because it's such an essential part of the workflow. Changes in sync protocol need to be carefully coordinated and tested between desktop and mobile so I think it makes sense to have them closely linked.

**Comment by okdistribute** [comment]
I don't think I've ever seen a mobile and desktop app in the same repository. Do you have any examples of that?

Would a really big git repo become sluggish over time? 

I'm not sure a monorepo will make it easier over time - especially for new folks to enter a particular project. 

Either way monorepo or not, updates will need to be closely coordinated between applications and modules. 

The examples you give of material-ui and react-router are not mixing heavy dependencies between the various packages (electron and react native and leveldb and node servers). 

I'd say it's more work than it's worth, with a small team and volunteers.

**Comment by okdistribute** [comment]
I've seen people use https://github.com/greenkeeperio/greenkeeper for keeping repository dependencies up to date in npm world. I've used it for years and it is pretty great.

**Comment by gmaclennan** [comment]
@karissa Yeah that's been my feeling in the past too. However, the "sell" of monorepos is to reduce the work. I think there is an inflection point where the work of maintaining all these modules becomes more work than setting up a monorepo. Greenkeeper would work for some cases, but two key workflows I'm thinking about which are currently really hard:

1. We start working on changes to the sync protocol. We need to test these changes in the apps and make sure they don't break sync between Mapeo Desktop and Mapeo Mobile. The dependency tree is something like `mapeo-mobile -> osm-p2p -> mapeo-server -> mapeo-core`. If we publish a non-major bump in mapeo-core we can't really control the moment it starts getting used in mapeo-mobile. `package-lock` doesn't really help because (a) it causes lots of problems of its own and (b) it only really works for direct deps (e.g. depth=1), and many important modules are not direct deps of mapeo-mobile. Testing without publishing means creating many `npm links` locally, and also having the mobile dev env setup. Every npm install wipes out npm links, so they need to be set up manually every time. Testing with CI would mean publishing betas of `mapeo-core` plus all the intermediate packages between mapeo-core and mapeo-mobile, and then greenkeeper could run to check the changes work, and then we could change all the betas and re-publish.

2. We move shared code from mapeo mobile and mapeo desktop into a separate module, e.g. a shared component or the API code in the frontend. Developing this requires testing and visualizing how it looks on both desktop and mobile. This either requires a complex `npm link` setup (which always breaks for me) or lots of npm publishing, which can accidentally break builds, or a really strict beta-track publishing on npm. Currently my solution to this is just to copy and paste the same code in both mapeo mobile and desktop, which is not ideal.

I'm basically looking for a way to simplify the work needed in these two workflows, and monorepos seem like _one_ solution to this but I'm not sure what other options are there. I've read a few articles recommending them not just for big projects with lots of maintainers, but also for smaller projects.

The additional benefit is consolidating issues in a single place, but that also comes with downsides. Currently I know the different related modules fairly well, but I still find it hard to keep track of issues across so many repos. There might be different solutions for this too.

I haven't seen an example of combining a desktop and mobile in the same repo, but there is no reason why it shouldn't work, might be worth testing out.

**Comment by okdistribute** [comment]
I was thinking of putting osm p2p as a dependency to mapeo core, and then mapeo server instantiating core for you, which would fix that issue. I never got around to it though I think it's not necessary for consumers of mapeo server to have to pass osm p2p down the pipe like that

**Comment by okdistribute** [comment]
As for sharing mapeo mobile and mapeo desktop code, I've seen people in react community create something like "APPNAME-components" which has any shared modules, then required upstream by the apps

**Comment by gmaclennan** [comment]
Some more thoughts on this:

- Having Mapeo Mobile and Mapeo Desktop in a single repo could help with issue reporting, it would be a single place for users to report issues with the apps, especially when issues are relating to sync and affect both apps.
- Including non-user-facing components in the same repo could result in too many issues and mix user bugs with internal upstream bugs.
- Expo has [started supporting electron](https://dev.to/evanbacon/making-desktop-apps-with-electron-react-native-and-expo-5e36), e.g. building mobile, desktop and web apps from the same codebase. I think combining desktop and mobile in the same repo is possible, but keeping them each with their own `node_modules` will be important to keep bundle-size low, because electron currently packages everything which is in the apps `node_modules`.
- There are two aspects to a monorepo: (1) keeping all the packages in the same git repo to ease development (e.g. no need to publish-install a dep to test changes) and (2) managing publishing of related packages. (1) can be done with yarn workspaces (easier to set up) and (2) needs lerna (lots to lern to get this set up). I see (1) as simplifying the setup of setting up `npm link` between packages and managing multiple peer dependencies, particularly with React.
- [Yarn 2 has been released](https://dev.to/arcanis/introducing-yarn-2-4eh1) and Yarn 1 switched to maintenance mode. It completely changes how packages are managed in a way that is incompatible with npm, and does not work with react-native. It has made me wary of using Yarn at all.
- I think Lerna could be too much tooling to deal with at this stage.
- [`npm@7`](https://blog.npmjs.org/post/186983646370/npm-cli-roadmap-summer-2019) will support something similar to yarn workspaces, and should be released in the next few months.

So maybe... only include MD and MM and shared UI components in the same repo? And wait for npm with workspaces support?

**Comment by okdistribute** [comment]
> Having Mapeo Mobile and Mapeo Desktop in a single repo could help with issue reporting, it would be a single place for users to report issues with the apps, especially when issues are relating to sync and affect both apps.
As someone who enjoys a clean issue tracker and has spent many years as the one pruning them, I don't think having Mobile and Desktop in the same repo will help issue reporting, I think it'll actually do the opposite. There are many features in Desktop that are desktop-only and then we'll have to organize and manage tags. If you want to have cross-project issue trackers there are other solutions for that, either a "help desk" or a organizational 'project'


> There are two aspects to a monorepo: (1) keeping all the packages in the same git repo to ease development (e.g. no need to publish-install a dep to test changes) and (2) managing publishing of related packages. (1) can be done with yarn workspaces (easier to set up) and (2) needs lerna (lots to lern to get this set up). I see (1) as simplifying the setup of setting up npm link between packages and managing multiple peer dependencies, particularly with React.
The only module we have that seems to have this `npm link` issue is react-mapfilter, is that right? Publish-installing a single dep with a particular test suite should run much faster than trying to run the entire repository's test suite (presumably with mobile, desktop, other components potentially?), unless those tests & builds are relatively quick, so I see the publish-install as a feature rather than a bug (also npm link works exactly like npm install would).

**Comment by okdistribute** [comment]
There is a pull request to turn `mapeo-core` into a mono repo, captured in https://github.com/digidem/mapeo-rfc/pull/4
